Writing style updates to Developing dynamic inventory topic (#71245)
ansible/ansible: docs/docsite/rst/dev_guide/developing_inventory.rst

Developing dynamic inventory
****************************

Ansible can pull inventory information from dynamic sources, including cloud sources, by using the supplied :ref:`inventory plugins <inventory_plugins>`. For details about how to pull inventory information, see :ref:`dynamic_inventory`. If the source you want is not currently covered by existing plugins, you can create your own inventory plugin as with any other plugin type.

In previous versions, you had to create a script or program that could output JSON in the correct format when invoked with the proper arguments.
You can still use and write inventory scripts, as we ensured backwards compatibility via the :ref:`script inventory plugin <script_inventory>`
and there is no restriction on the programming language used.
If you choose to write a script, however, you will need to implement some features yourself such as caching, configuration management, dynamic variable and group composition, and so on.
If you use :ref:`inventory plugins <inventory_plugins>` instead, you can leverage the Ansible codebase and add these common features automatically.

.. contents:: Topics
   :local:

.. _inventory_sources:

> What are inventory sources?

Reviewer: I'd undo this change.

.. _developing_inventory_inventory_plugins:

> What are inventory plugins?

Reviewer: Here too, I'd avoid using a question for the header.

===========================

Like most plugin types (except modules), inventory plugins must be developed in Python. They execute on the controller and should therefore adhere to the :ref:`control_node_requirements`.

Most of the documentation in :ref:`developing_plugins` also applies here. You should read that document first for a general understanding and then come back to this document for specifics on inventory plugins.

Normally, inventory plugins are executed at the start of a run, and before the playbooks, plays, or roles are loaded.
However, you can use the ``meta: refresh_inventory`` task to clear the current inventory and execute the inventory plugins again, and this task will generate a new inventory.

> If you use the persistent cache, inventory plugins can also use the configured cache plugin to store and retrieve data. This data retrieval avoids making repeated and costly external calls.

Reviewer: ```suggestion
If you use the persistent cache, inventory plugins can also use the configured cache plugin to store and retrieve data. Caching inventory avoids making repeated and costly external calls.
```

        self.inventory = inventory
        self.templar = Templar(loader=loader)

> It is upto the plugin now to parse the provided inventory source and translate it to Ansible inventory.

Reviewer: Typo!
```suggestion
It is up to the plugin now to parse the provided inventory source and translate it into Ansible inventory.
```

  - ``update_cache_if_changed`` sets the cache plugin only if ``self._cache`` has been modified, before the ``parse`` method completes
  - ``clear_cache`` flushes the cache, ultimately by calling the cache plugin's ``flush()`` method, whose implementation is dependent upon the particular cache plugin in use. Note that if the user is using the same cache backend for facts and inventory, both will get flushed. To avoid this, the user can specify a distinct cache backend in their inventory plugin configuration.

.. _inventory_source_common_format:

Common format for inventory sources
-----------------------------------

To simplify development, most plugins use a standard YAML-based configuration file as the inventory source. The file has only one required field ``plugin``, which should contain the name of the plugin that is expected to consume the file.
Depending on other common features used, you might need other fields, and you can add custom options in each plugin as required.
For example, if you use the integrated caching, ``cache_plugin``, ``cache_timeout`` and other cache-related fields could be present.

.. _inventory_development_auto:

The 'auto' plugin
-----------------

> From Ansible 2.5 onwards, we include the :ref:`auto inventory plugin <auto_inventory>` that is enabled by default. This plugin loads other plugins that use the common YAML configuration format and specifies a ``plugin`` field with a matching inventory plugin name. The 'auto' plugin makes it easier to use your plugin without having to update configurations.

Reviewer: ```suggestion
From Ansible 2.5 onwards, we include the :ref:`auto inventory plugin <auto_inventory>` and enable it by default. If the ``plugin`` field in your standard configuration file matches the name of your inventory plugin, the ``auto`` inventory plugin will load your plugin. The 'auto' plugin makes it easier to use your plugin without having to update configurations.
```

    }

If any of the elements of a group are empty, they may be omitted from the output.

When called with the argument ``--host <hostname>`` (where <hostname> is a host from above), the script must print either an empty JSON hash/dictionary, or a hash/dictionary of variables to make them available to templates and playbooks. For example::

    {
        "VAR001": "VALUE",
        "VAR002": "VALUE",
    }

> Printing variables is optional. If the script does not do it, the script should print an empty hash or dictionary.

Reviewer: ```suggestion
Printing variables is optional. If the script does not print variables, it should print an empty hash or dictionary.
```
